### PR TITLE
Get rid of compiler warnings; python3 for scalapack and external distributions included in github self-tests 

### PR DESCRIPTION
**Pull request by MatthiasHeilManchester**
Get rid of compiler warnings, mainly arising from unused variables.

Mainly straightforward. There were a couple of instances where switch statements in superlu_dist.c didn't provide a default option. Initially provided a broken default, but two of those were triggered, legally, it seems, when running the code. Seems more a problem of slight cowboy code since, by default, nothing needed to be done (so the code did the right thing but the compiler was right to complain!). Left the other instances there (with diagnostics preceding the abort so we can decide on a case-by-case basis what needs to be done.  (Note that this code is basically ported from superlu, so not really ours.

Note: Changes include changes also made by @wobblygomboc in his pending pull request (which still needs to some attention before it can be merged). Resulting conflicts (if any) should be easy to resolve.

Note: Tests pass cleanly on my laptop with paranoia and range checking.

While revising this pull request I (and Puneet!) also upgraded python scripts for scalapack build to python3. The converted scripts now live in a separate directory and get copied into the scalapack directory after the tar file has been unpacked.

Also included the build (and associated self-tests) for the third-party libraries into the ubuntu self-test action (for mpi only).

### CODE REVIEW

**Comment by puneetmatharu** [comment]
I'm happy to merge this unless you think there is anything that really needs a look at(?). If you want me to be picky, you could delete the code that you commented out in my space-time elements. :)

**Comment by MatthiasHeilManchester** [comment]
> I'm happy to merge this unless you think there is anything that really needs a look at(?). If you want me to be picky, you could delete the code that you commented out in my space-time elements. :)

yeah, actually I will do this. Meant to run through all the permutations again on my work computer to see if the warnings are really gone. I'll let you know when I'm done, then we can declare this to be the release candidate.

**Comment by MatthiasHeilManchester** [comment]
@puneetmatharu Ready for you! Self tests (incl. mpi test of the external distributions) all pass and documentation builds ok.

File: config/configure_options/private_configure_options/full_opt_all_libraries_mpi [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+#--with-pord=/home/mheil/local/mumps_and_scalapack_default_installation/lib/libpord.a 
+#--with-mumps=/home/mheil/local/mumps_and_scalapack_default_installation
+#--with-blas=/home/mheil/local/lib/blas/blas.a
+#--with-lapack=/home/mheil/local/lib/lapack/lapack.a
+#--with-hypre=/home/mheil/local/hypre_default_installation_mpi
+#--with-trilinos=/home/mheil/local/trilinos_default_installation_mpi
+
+
+
+#--with-boost=/home/mheil/version_from_christians_first_checkin/external_distributions/boost/boost_default_installation
```
Comment (puneetmatharu) [inline config/configure_options/private_configure_options/full_opt_all_libraries_mpi:125]: Also could clean this up (optional)

File: external_distributions/mumps_and_scalapack/Makefile.am [hunk 2]
```diff
@@ -127,13 +127,16 @@ endif
 	echo "#Install with blacs (downloaded on the fly)" >> build_script.bash
 	echo "echo \"Bypassing test of executables (because it breaks on hector, or any\"">> build_script.bash
 	echo "echo \"other machine where we cross compile)\"">> build_script.bash
-	echo "cp script/framework.py script/framework.py.orig" >> build_script.bash
-	echo "sed 's/.\/tmpf/date/g' script/framework.py.orig | sed 's/.\/tmpc/date/g' > script/framework.py ">> build_script.bash
-	echo "patch script/blacs.py < ../../build_scripts/blacs.py.patch  ">> build_script.bash
+#OLD (PYTHON2)	echo "cp script/framework.py script/framework.py.orig" >> build_script.bash
+	echo "cp ../../scalapack_python3_scripts/*py script" >> build_script.bash
+#OLD (PYTHON2)	echo "sed 's/.\/tmpf/date/g' script/framework.py.orig | sed 's/.\/tmpc/date/g' > script/framework.py ">> build_script.bash
+#OLD (PYTHON2)	echo "patch script/blacs.py < ../../build_scripts/blacs.py.patch  ">> build_script.bash
```
Comment (puneetmatharu) [inline external_distributions/mumps_and_scalapack/Makefile.am:133]: Might as well kill the lines with `#OLD (PYTHON2)`, we can always flick back through them with Git.

File: external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ mangling, so adding debug statements to the python scripts can have
+ unintended consequences (as I found out).
+
+  To compare the modified files against the original ones:
+
+    cd external_distributions/mumps_and_scalapack/scalapack_python3_scripts/
+    mkdir junk
+    cp scalapack_installer.tgz junk/
+    cd junk/
+    tar xvfz scalapack_installer.tgz
```
Comment (puneetmatharu) [inline external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README:17]: Just a little tip (that I only recently stumbled onto); you can just specify the location that files are untarred to with the `-C` flag. Therefore, from the outer directory, you could just call:
```bash
tar xvfz scalapack_installer.tgz -C junk/
```

File: external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  To compare the modified files against the original ones:
+
+    cd external_distributions/mumps_and_scalapack/scalapack_python3_scripts/
+    mkdir junk
+    cp scalapack_installer.tgz junk/
+    cd junk/
+    tar xvfz scalapack_installer.tgz 
+    cd scalapack_installer_0.94/script/
+    find . -name '*.py' -exec meld {} ../../../scalapack_python3_scripts/{} \;
```
Comment (puneetmatharu) [inline external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README:19]: Assumes people have `meld` -- I don't. (IIRC, there's no support for it on macOS.) [As an alternative, you can replace `meld` with `git diff --no-index --`](https://stackoverflow.com/a/15110387).

File: external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+  Note that test are bypassed to allow for cross-compilation (where OS on
```
Comment (puneetmatharu) [inline external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README:21]: "test" -> "tests"
"where OS" -> "where the OS"

File: src/generic/superlu_dist.c [hunk 6]
```diff
@@ -182,25 +182,25 @@ void superlu_dist_distributed_matrix(int opt_flag, int allow_permutations,
   int_t *perm_r; /* row permutations from partial pivoting */
   int_t *perm_c; /* column permutation vector */
   int_t *etree;  /* elimination tree */
-  int_t *rowptr, *colind;  /* Local A in NR*/
+  int_t *rowptr=0; int_t *colind;  /* Local A in NR*/
```
Comment (puneetmatharu) [inline src/generic/superlu_dist.c:185]: Might be worth initialising both of this pointers using `NULL`, i.e. use
```c
int_t *rowptr=NULL; 
int_t *colind=NULL;
```

File: src/generic/superlu_dist.c [hunk 7]
```diff
@@ -182,25 +182,25 @@ void superlu_dist_distributed_matrix(int opt_flag, int allow_permutations,
...
   int_t job, rowequ, colequ, iinfo, need_value, i, j, irow, icol;
-  int_t m_loc, fst_row, nnz, nnz_loc, dist_mem_use;
+  int_t m_loc, fst_row, nnz, nnz_loc; /* dist_mem_use; */
   int_t *colptr, *rowind;
   NRformat_loc *Astore;
   SuperMatrix GA;      /* Global A in NC format */
   NCformat *GAstore;
-  double *a_GA;
+  double *a_GA=0;
```
Comment (puneetmatharu) [inline src/generic/superlu_dist.c:192]: `NULL`?

File: src/generic/superlu_dist.c [hunk 8]
```diff
@@ -182,25 +182,25 @@ void superlu_dist_distributed_matrix(int opt_flag, int allow_permutations,
...
   NCPformat *GACstore;
   Glu_persist_t *Glu_persist;
-  Glu_freeable_t *Glu_freeable;
+  Glu_freeable_t *Glu_freeable=0;
 
   /* Other stuff needed by SuperLU */
-  double  *berr;
-  double *a, *X, *b_col;
+  double  *berr=0;
+  double *a=0; double *X, *b_col;
```
Comment (puneetmatharu) [inline src/generic/superlu_dist.c:200]: Same as above.

File: src/generic/superlu_dist.c [hunk 9]
```diff
@@ -1256,21 +1266,21 @@ void superlu_dist_global_matrix(int opt_flag, int allow_permutations,
...
+  int_t Equil, factored, notran, permc_spec; /*, dist_mem_use; */
   NCformat *Astore;
   NCPformat *ACstore;
   Glu_persist_t *Glu_persist;
-  Glu_freeable_t *Glu_freeable;
+  Glu_freeable_t *Glu_freeable=0;
 
   /* Other stuff needed by SuperLU */
-  double  *berr;
+  double  *berr=0;
```
Comment (puneetmatharu) [inline src/generic/superlu_dist.c:1279]: `NULL`?

**Review by puneetmatharu — CHANGES_REQUESTED** [review]
Looks good to me for the most part! I've added a few small remarks on possible changes

File: config/configure_options/private_configure_options/paranoid_range_check_all_libraries_mpi [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+#--with-pord=/home/mheil/local/mumps_and_scalapack_default_installation/lib/libpord.a 
+#--with-mumps=/home/mheil/local/mumps_and_scalapack_default_installation
+#--with-blas=/home/mheil/local/lib/blas/blas.a
+#--with-lapack=/home/mheil/local/lib/lapack/lapack.a
+#--with-hypre=/home/mheil/local/hypre_default_installation_mpi
+#--with-trilinos=/home/mheil/local/trilinos_default_installation_mpi
+
+
+
+#--with-boost=/home/mheil/version_from_christians_first_checkin/external_distributions/boost/boost_default_installation
```
[the remark this answers is not in the document]
Comment (MatthiasHeilManchester) [reply config/configure_options/private_configure_options/paranoid_range_check_all_libraries_mpi:125]: Prefer to keep these (only partially out of laziness; it shows what other options exist and I often cut and paste from such abandoned parts of the config file). And this one lives in private_configure_options anyway.

File: config/configure_options/private_configure_options/full_opt_all_libraries_mpi (hunk 1 above)
Comment (MatthiasHeilManchester) [reply config/configure_options/private_configure_options/full_opt_all_libraries_mpi:125]: Ditto

File: external_distributions/mumps_and_scalapack/Makefile.am (hunk 2 above)
Comment (MatthiasHeilManchester) [reply external_distributions/mumps_and_scalapack/Makefile.am:133]: If you don't mind, I'll keep this in. It's true that flicking back through the history is possible but the way this is being customised/built is not straightforward and this comment will remind me (without having to consult extra information) about the logic. In the current state I'm likely to be the only one to go one there if stuff breaks again, and hopefully it'll all be superseded by lightning fast cmake installs anyway and they're (hopefully) bypassing all this horror.

File: external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README (hunk 3 above)
Comment (MatthiasHeilManchester) [reply external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README:17]: Thanks, but the code above is more transparent (Am I defensive today? Must act on at least one of your suggestions!). Let's see...

File: external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README (hunk 4 above)
Comment (MatthiasHeilManchester) [reply external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README:19]: Done (though I'm not convinced git is quite the replacement; suggested using diff or sdiff).

File: external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README (hunk 5 above)
Comment (MatthiasHeilManchester) [reply external_distributions/mumps_and_scalapack/scalapack_python3_scripts/README:21]: Done. Also added a closing bracket! :)

File: src/generic/superlu_dist.c (hunk 8 above)
Comment (MatthiasHeilManchester) [reply src/generic/superlu_dist.c:200]: Done.

File: src/generic/superlu_dist.c (hunk 7 above)
Comment (MatthiasHeilManchester) [reply src/generic/superlu_dist.c:192]: Done

File: src/generic/superlu_dist.c (hunk 9 above)
Comment (MatthiasHeilManchester) [reply src/generic/superlu_dist.c:1279]: Done (also searched for other "=0" patterns and replaced a few more.

Done.

**Comment by MatthiasHeilManchester** [comment]
Dealt with all of these and added comments while making changes on my local machine. Will let you know when they've passed on github then you can have another look.

**Comment by puneetmatharu** [comment]
I'm happy with these changes -- will merge it in now!

**Pull request merged**